### ISSUE TITLE
Type hints doesn't work as expected, because they aren't present in dask source code.

### ISSUE THREAD

**Issue by karolzlot** [dask/dask#8853]
**What happened**:

I'm trying to make a function which accepts dask.dataframe. Type hints doesn't work as expected however. If I switch back to pandas, then type hints works ok.

**What you expected to happen**:

No red underlines in vscode indicating errors.

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe as dd

def func1(
    df: dd.DataFrame
) -> dd.DataFrame:

    df = df.set_index("aaa")

    df = df.rename(columns=lambda x: str(x).strip())

    return df
```
![image](https://user-images.githubusercontent.com/11590960/160240465-79fb13a0-3326-4da3-8b42-d70c8696cfae.png)

![image](https://user-images.githubusercontent.com/11590960/160240462-bbe5d7c1-f7d4-414c-bedb-edccb9c6e0d1.png)

**Anything else we need to know?**:

I use vscode and:

![image](https://user-images.githubusercontent.com/11590960/160240698-b92e733f-081e-43ba-812a-2facfe8f6883.png)

It seems that dask (unlike pandas) doesn't have type hints defined, and they are generated by pylance...

**Environment**:

- Dask version: 2022.3.0
- Python version: 3.10.4
- Operating System: W10
- Install method (conda, pip, source): conda

**Comment by pavithraes** [comment]
Hi @karolzlot, welcome to Dask, thanks for opening this issue! I'm not able to reproduce this (I'm on a macOS system, but I matched your Python and Dask versions). Could you please share more information about the extensions you're using and the `Python Language Server` (I'm assuming it's Pylance, but I can't reproduce using Jedi either). It may be a Windows thing?

It also seems like this is more of a usage question than a bug report or feature request. We encourage people with such questions to ask them at the [Dask Discourse](https://dask.discourse.group/). Would you mind opening a discussion topic about this over there? If you do so, feel free to close this issue and we will continue the conversation on discourse!

**Comment by karolzlot** [comment]
@pavithraes 
Yes, I'm using pylance.

I'm not sure why you can't reproduce it.

You need to enable "Use Library Code For Types" setting and then you should be able to see those red underlines.

It's not about usage, it's a feature request. I don't see type hints in source code and this issue is about adding them.

**Comment by bryanwweber** [comment]
This is a known issue that is being worked on, see https://github.com/dask/dask/pull/8706

**Comment by d-v-b** [comment]
+1 for integrating type hints with dask. `dask.delayed` seems like an obvious target for this; beyond better integration with the increasingly typed python ecosystem, you could parse function annotations to get the information used in the [`nout` keyword argument](https://github.com/dask/dask/blob/06a6b5afe120a2545f5496d2102516361494ba48/dask/delayed.py#L299)

**Comment by ViktorSky** [comment]
> Hi @karolzlot, welcome to Dask, thanks for opening this issue! I'm not able to reproduce this (I'm on a macOS system, but I matched your Python and Dask versions). Could you please share more information about the extensions you're using and the `Python Language Server` (I'm assuming it's Pylance, but I can't reproduce using Jedi either). It may be a Windows thing?
> 
> It also seems like this is more of a usage question than a bug report or feature request. We encourage people with such questions to ask them at the [Dask Discourse](https://dask.discourse.group/). Would you mind opening a discussion topic about this over there? If you do so, feel free to close this issue and we will continue the conversation on discourse!

Try using vscode:
extensions: Pylance (Microsoft), Mypi Type Checker (Microsoft)
config: Type Checking Mode: strict
![Uploading extensions.png…]()
![Uploading config.png…]()

**Comment by C-Loftus** [comment]
Is there a recommendation for how to work within VSCode since there is no intellisense or typing? Or has anyone recently gotten VSCode intellisense to work recently with Dask? Makes it difficult to know what is able to be called without running into runtime errors. 

I am a bit confused since there seems to be mypy support https://github.com/dask/dask/pull/8706 but there is no intellisense. Not sure if I am missing something.

Even if I import from the module or make sure pylance uses library code it does not work. 

Corresponding pylance issue says https://github.com/microsoft/pylance-release/issues/4801#issuecomment-1708900529 

![image](https://github.com/user-attachments/assets/0981d9ca-06d3-44f5-be39-3cd28cebe078)
    
![image](https://github.com/user-attachments/assets/66d2dfc8-1b29-43cd-a76f-7f2d80eda89f)

**Comment by ViktorSky** [comment]
> ¿Existe alguna recomendación sobre cómo trabajar dentro de VSCode ya que no hay intellisense o typing? ¿O alguien ha conseguido recientemente VSCode intellisense para trabajar recientemente con Dask? Hace que sea difícil saber cómo se puede llamar sin encontrarse con errores de tiempo de ejecución.
> 
> Estoy un poco confundido ya que parece haber soporte mypy #8706 pero no hay inteligencia. No estoy seguro si me falta algo.
> 
> Incluso si importo desde el módulo o me aseguro de que pylance use código de biblioteca, no funciona.
> 
> El problema de pilance correspondiente dice [microsoft/pylance-release#4801 (comentario)](https://github.com/microsoft/pylance-release/issues/4801#issuecomment-1708900529)
> 
> ![imagen](https://private-user-images.githubusercontent.com/70598503/392489216-0981d9ca-06d3-44f5-be39-3cd28cebe078.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzNDI3NzgsIm5iZiI6MTczMzM0MjQ3OCwicGF0aCI6Ii83MDU5ODUwMy8zOTI0ODkyMTYtMDk4MWQ5Y2EtMDZkMy00NGY1LWJlMzktM2NkMjhjZWJlMDc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA0VDIwMDExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmNmYzNTU2ZTMzZDY1MGJkMDBlM2ZiNTY1NzhiYTk0MTI0ODQwODI3MTY3MDc4YjcyMDZmZDI0NGY3MzlmNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oBZsjIVHAX5s7iI_UvQiT3ZHbaF8neWnBbcLPaIAEKM)
> 
> ![imagen](https://private-user-images.githubusercontent.com/70598503/392490124-66d2dfc8-1b29-43cd-a76f-7f2d80eda89f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzNDI3NzgsIm5iZiI6MTczMzM0MjQ3OCwicGF0aCI6Ii83MDU5ODUwMy8zOTI0OTAxMjQtNjZkMmRmYzgtMWIyOS00M2NkLWE3NmYtN2YyZDgwZWRhODlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA0VDIwMDExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3NjQzZTEyODA1NjFiOWQ3NDM1NWNjMzlhYTY5Mzg5YzVhODIyNzBmOTBlYzQ5M2UyMjQ3MTg2ZGNkNzVhNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iVZLvVktj414UkkeTpiXwW1Q_yr9WYWvefFH_J3SHps)

You can try this configuration until they release an update:

`.vscode/setting.json`
```json
{
    "python.analysis.typeCheckingMode": "standard",
    "python.analysis.typeEvaluation.analyzeUnannotatedFunctions": false
}
```

Use absolute imports for functions imported from dask

```python
from dask.dataframe.core import DataFrame

def func1(df: DataFrame) -> DataFrame:
    df = df.set_index("aaa")
    df = df.rename(columns=lambda x: str(x).strip())
    return df
```

**Comment by flying-sheep** [comment]
> Use absolute imports for symbols imported from dask

There is a discrepancy between type hints and documentation.
- in the docs, the import path of `Array` is `dask.array.Array`
- according to the type hints, it’s `dask.array.core.Array`

@pavithraes can you confirm that both are stable and supported import locations?

**Comment by phofl** [comment]
>can you confirm that both are stable and supported import locations?

No, they are not. Please only import from dask.array directly, not from core

**Comment by flying-sheep** [comment]
OK, so what’s your take on https://github.com/microsoft/pylance-release/issues/4801#issuecomment-1708900529?

I disagree with them about the `from ... import <name> as <name>` pattern, I’m pretty sure they invented that.

But they do have a point that you don’t signify an export. Is there a chance we can get an `__all__` defined in all the API packages?

**Comment by phofl** [comment]
We added __all__ in a few places a little while ago. You can open a quick PR if it's missing somewhere you are interested in, happy to review

**Comment by phofl** [comment]
sounds good! But please split this into 2 PRs, one adding the annotations and a second one adding a check

**Comment by flying-sheep** [comment]
OK, apparently this pattern is a thing: https://typing.readthedocs.io/en/latest/spec/distributing.html#library-interface-public-and-private-symbols

Would you prefer this style to `__all__`?

**Comment by ViktorSky** [comment]
> >can you confirm that both are stable and supported import locations?
> 
> No, they are not. Please only import from dask.array directly, not from core

The problem with doing that is that type checkers think that importable objects are only for internal use and not for external use.

That can be solved by using an alias. 

```py
from dask.array.core import Array as Array
```

If you want to know more about these particular cases, you should understand how type checkers read stub files (`.pyi`).

**Comment by flying-sheep** [comment]
Yes, that’s what we’ve been talking about the last few comments, including how to fix it.
